### PR TITLE
[CSPL-2699]: adding Azure and GCP bucket access using sdk

### PR DESCRIPTION
**Pull request by vivekr-splunk**
Signed-off-by: Vivek Reddy <vivekrsplunk@github.com>

The Splunk Operator for Kubernetes currently supports AWS S3 buckets, allowing users to integrate Splunk with AWS storage seamlessly. However, customers utilizing Google Cloud Platform (GCP) and Microsoft Azure are required to use interoperability solutions to access their respective Blob storage services. This pull request introduces native support for GCP storage buckets and enhances Azure integration by implementing Azure AD authentication alongside traditional credential methods.

### Key Enhancements:

- **GCP Storage Support:**
  - **GCP SDK Integration:** Introduced the Google Cloud Platform (GCP) SDK to enable direct interaction with GCP storage buckets.
  - **GCP SDK Wrapper:** Developed a wrapper for the GCP SDK, mirroring the existing AWS S3 client wrapper, to facilitate operations such as listing and downloading objects from GCP storage buckets.
  
- **Azure Storage Enhancements:**
  - **Azure AD Authentication:** Implemented Azure Active Directory (Azure AD) authentication to provide secure and streamlined access to Azure Blob Storage.
  - **Azure SDK Integration:** Leveraged the Azure SDK to manage credentials and handle Blob storage operations effectively.

## Motivation

Expanding the Splunk Operator's storage capabilities to include GCP and Azure addresses the needs of a broader user base operating in multi-cloud environments. Native support eliminates the complexities and limitations associated with interoperability solutions, offering users a more robust, secure, and efficient way to manage their storage integrations within Splunk.

### GCP Integration

1. **GCP SDK Wrapper:**
   - Created a new module `gcpbucketclient` that encapsulates GCP SDK functionalities.
   - Implemented methods for:
     - Listing objects in GCP storage buckets.
     - Downloading objects from GCP storage buckets.

2. **Configuration Updates:**
   - Added configuration options in the operator's settings to specify GCP credentials and bucket details.
   - Updated Kubernetes `ServiceAccount` and `Role` definitions to accommodate GCP access permissions.

3. **Documentation:**
   - Updated README and user guides to include instructions for configuring GCP storage integration.

### Azure Enhancements

1. **Azure AD Authentication:**
   - Integrated Azure AD authentication mechanisms to secure access to Azure Blob Storage.
   - Configured the operator to use Azure AD tokens for authenticating storage operations.

2. **Azure SDK Integration:**
   - Utilized the Azure SDK to handle credential management and Blob storage interactions.
   - Ensured compatibility with existing Azure storage configurations and workflows.

3. **Configuration Updates:**
   - Enhanced the operator's configuration to support Azure AD authentication parameters.
   - Updated Kubernetes manifests to include necessary Azure RBAC roles and permissions.

4. **Documentation:**
   - Expanded documentation to cover Azure AD authentication setup and troubleshooting steps.

## Testing

- **Unit Tests:**
  - Added unit tests for the new GCP SDK wrapper to ensure reliable interactions with GCP storage.
  - Enhanced existing tests to cover Azure AD authentication flows.

- **Integration Tests:**
  - Performed end-to-end testing with GCP storage buckets to validate listing and downloading functionalities.
  - Verified Azure Blob Storage access using Azure AD authentication, ensuring secure and authorized operations.

- **Regression Testing:**
  - Ensured that existing AWS S3 functionalities remain unaffected by the new integrations.
  - Conducted performance testing to confirm that the operator maintains optimal performance with multi-cloud support.

## Impact

This enhancement broadens the Splunk Operator's compatibility, making it a more versatile tool for organizations leveraging multi-cloud strategies. By providing native support for GCP and Azure storage services, users can manage their Splunk integrations more effectively, reducing dependency on third-party interoperability solutions and simplifying their infrastructure management.

## Related Issues

- **Issue CSPL-2824:** Support for GCP Storage Buckets
- **Issue CSPL-2825:** Implement Azure AD Authentication for Blob Storage Access

## Future Work

- **GCP Storage Operations:** Extend GCP storage support to include uploading objects and managing bucket configurations.
- **Azure Storage Features:** Incorporate additional Azure Blob Storage functionalities such as container management and advanced access controls.
- **Multi-Cloud Optimization:** Explore optimizations for simultaneous multi-cloud storage operations to enhance performance and reliability.
- **User Interface Enhancements:** Update the operator's UI/dashboard to provide better visibility and control over multi-cloud storage integrations.

## Checklist

- [x] Code follows the project's coding standards
- [x] Relevant documentation has been updated
- [x] All new and existing tests pass
- [x] Changes have been tested in a development environment

### CODE REVIEW

File: .github/workflows/int-test-gcp-workflow.yml [hunk 1]
```diff
@@ -0,0 +1,287 @@
+name: Integration Test on GCP Workflow
+
+on:
+  push:
+    branches:
+      - develop
+      - main
+      - google-object-storage-support
```
Comment (rlieberman-splunk) [inline .github/workflows/int-test-gcp-workflow.yml:8]: Before merging, remove this branch from the trigger.

**Review by rlieberman-splunk — APPROVED** [review]

File: docs/AppFramework.md [hunk 2]
```diff
@@ -622,91 +752,843 @@ The App Framework does not preview, analyze, verify versions, or enable Splunk A
...
+
+    ```bash
+    az role assignment create --assignee "f0f04120-6a36-49bc--**************" --role 'Storage Blob Data Reader' --scope /subscriptions/f428689e-c379-4712--**************/resourceGroups/splunkOperatorResourceGroup/providers/Microsoft.Storage/storageAccounts/mystorageaccount
+    ```
+
+    After this command, you can use the App Framework for Azure Blob without secrets.
+
+### **Azure Blob Authorization Recommendations:**
+
+- **Granular Access:** Azure allows "Managed Identities" assignment at the "storage accounts" level as well as at specific containers (buckets) levels. A managed identity assigned read permissions at a storage account level will have read access for all containers within that storage account. As a good security practice, assign the managed identity to only the specific containers it needs to access, rather than the entire storage account.
```
Comment (akondur) [inline docs/AppFramework.md:848]: Can we mark "Managed Identities" and "storage accounts" in bold?

File: docs/AppFramework.md [hunk 3]
```diff
@@ -622,91 +752,843 @@ The App Framework does not preview, analyze, verify versions, or enable Splunk A
...
+  
+- **Avoid Shared Access Keys:** In contrast to "Managed Identities", Azure allows "shared access keys" configurable only at the storage accounts level. When using the `secretRef` configuration in the CRD, the underlying secret key will allow both read and write access to the storage account (and all containers within it). Based on your security needs, consider using "Managed Identities" instead of secrets. Additionally, there's no automated way to rotate the secret key, so if you're using these keys, rotate them regularly (e.g., every 90 days).
```
Comment (akondur) [inline docs/AppFramework.md:850]: Similar comment as above for everything in quotes.

File: pkg/splunk/enterprise/searchheadcluster.go [hunk 4]
```diff
@@ -270,6 +272,8 @@ var newSearchHeadClusterPodManager = func(client splcommon.ControllerClient, log
 
 // ApplyShcSecret checks if any of the search heads have a different shc_secret from namespace scoped secret and changes it
 func ApplyShcSecret(ctx context.Context, mgr *searchHeadClusterPodManager, replicas int32, podExecClient splutil.PodExecClientImpl) error {
+
```
Comment (akondur) [inline pkg/splunk/enterprise/searchheadcluster.go:275]: Nit: Can we remove these two empty lines?

File: pkg/splunk/enterprise/searchheadcluster.go [hunk 5]
```diff
@@ -432,6 +436,7 @@ func ApplyShcSecret(ctx context.Context, mgr *searchHeadClusterPodManager, repli
 
 // Update for searchHeadClusterPodManager handles all updates for a statefulset of search heads
 func (mgr *searchHeadClusterPodManager) Update(ctx context.Context, c splcommon.ControllerClient, statefulSet *appsv1.StatefulSet, desiredReplicas int32) (enterpriseApi.Phase, error) {
+
```
Comment (akondur) [inline pkg/splunk/enterprise/searchheadcluster.go:439]: Nit: Is there a reason why we have an empty line here?

File: pkg/splunk/enterprise/searchheadcluster.go [hunk 6]
```diff
@@ -465,6 +470,7 @@ func (mgr *searchHeadClusterPodManager) Update(ctx context.Context, c splcommon.
 
 // PrepareScaleDown for searchHeadClusterPodManager prepares search head pod to be removed via scale down event; it returns true when ready
 func (mgr *searchHeadClusterPodManager) PrepareScaleDown(ctx context.Context, n int32) (bool, error) {
+
```
Comment (akondur) [inline pkg/splunk/enterprise/searchheadcluster.go:473]: Same as above.

File: pkg/splunk/enterprise/searchheadcluster.go [hunk 7]
```diff
@@ -486,6 +492,7 @@ func (mgr *searchHeadClusterPodManager) PrepareScaleDown(ctx context.Context, n
 
 // PrepareRecycle for searchHeadClusterPodManager prepares search head pod to be recycled for updates; it returns true when ready
 func (mgr *searchHeadClusterPodManager) PrepareRecycle(ctx context.Context, n int32) (bool, error) {
+
```
Comment (akondur) [inline pkg/splunk/enterprise/searchheadcluster.go:495]: Same as above.

File: test/env.sh [hunk 8]
```diff
@@ -13,11 +13,11 @@
...
 : "${EKS_CLUSTER_K8_VERSION:=1.31}"
 # Below env variables required to run license master test cases
-: "${ENTERPRISE_LICENSE_S3_PATH:=}"
-: "${TEST_S3_BUCKET:=}"
+: "${ENTERPRISE_LICENSE_S3_PATH:=/test_licenses}"
+: "${TEST_S3_BUCKET:=splk-test-data-bucket}"
 # Below env variables required to run remote indexes test cases
-: "${INDEXES_S3_BUCKET:=}"
-: "${AWS_S3_REGION:=}"
+: "${INDEXES_S3_BUCKET:=splk-integration-test-bucket}"
```
Comment (akondur) [inline test/env.sh:19]: Are we hardcoding these for now? Not a bad idea

**Review by akondur — COMMENTED** [review]
Looks good. Minor comments

File: config/crd/bases/enterprise.splunk.com_clustermanagers.yaml [hunk 9]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.16.1
+    controller-gen.kubebuilder.io/version: (devel)
```
[the remark this answers is not in the document]
Comment (vivekr-splunk) [reply config/crd/bases/enterprise.splunk.com_clustermanagers.yaml:6]: this is auto generated code

File: test/env.sh (hunk 8 above)
Comment (vivekr-splunk) [reply test/env.sh:19]: yes

**Review by akondur — APPROVED** [review]

**Pull request merged**